iroh-resolver/src/unixfs_builder.rs

```diff
@@ -1,55 +1,172 @@
+            name: None,
+            entries: Default::default(),
+            typ: DirectoryType::Basic,
+            estimated_size: 0,
+            chunker: Chunker::fixed_size(),
+        }
+    }
 }
 
 impl DirectoryBuilder {
     pub fn new() -> Self {
-        Self::default()
+        Default::default()
     }
 
-    pub fn name(&mut self, name: impl Into<String>) -> &mut Self {
+    pub fn with_path(&mut self, path: PathBuf) -> &mut Self {
+        self.path = Some(path);
+        self
+    }
+
+    pub fn with_name(&mut self, name: impl Into<String>) -> &mut Self {
         self.name = Some(name.into());
         self
     }
 
+    pub fn add_dir(&mut self, dir: Directory) -> &mut Self {
+        let name = dir.name();
+        self.estimated_size += self.estimated_size + name.len() + CID_LEN;
+        if self.typ == DirectoryType::Basic
+            && self.estimated_size + BUFFER > self.chunker.chunk_size()
+        {
+            self.typ = DirectoryType::Hamt;
+        }
+        self.entries.push(Entry::Directory(dir));
+        self
+    }
+
     pub fn add_file(&mut self, file: File) -> &mut Self {
-        self.files.push(file);
+        let name = file.name();
+        self.estimated_size += self.estimated_size + name.len() + CID_LEN;
+        if self.typ == DirectoryType::Basic
+            && self.estimated_size + BUFFER > self.chunker.chunk_size()
+        {
+            self.typ = DirectoryType::Hamt;
+        }
+        self.entries.push(Entry::File(file));
         self
     }
 
-    pub async fn build(self) -> Result<Directory> {
-        let DirectoryBuilder { name, files } = self;
+    #[async_recursion(?Send)]
+    pub async fn build(mut self) -> Result<Directory> {
+        if let Some(path) = self.path {
+            let name = path
+                .file_name()
+                .and_then(|s| s.to_str())
+                .unwrap_or_default();
+
+            let mut builder = DirectoryBuilder::new();
+            builder.with_name(name);
+            let entries = std::fs::read_dir(path)?;
+            for entry in entries {
+                let entry = entry?;
+                let path = entry.path();
+                // TODO: this seems dumb, is this the only way to differentiate between
+                // files and dirs?
+                // TODO: what do we do with symlinks?
+                if path.is_file() {
+                    let mut file_builder = FileBuilder::new();
+                    file_builder.with_path(path);
+                    let f = file_builder.build().await?;
+                    builder.add_file(f);
+                } else if path.is_dir() {
+                    let mut dir_builder = DirectoryBuilder::new();
+                    dir_builder.with_path(path);
+                    let d = dir_builder.build().await?;
+                    builder.add_dir(d);
+                } else {
+                    anyhow::bail!("directory entry is neither file nor directory")
+                }
```

Comment:
reiterating my comments here:
this seems dumb, but when I poked around this seemed like that way to differentiate `FileTypes`

Also, not sure what I should be doing to add `symlinks`. Should I be adding a `struct SymlinkBuilder` & `struct Symlink`? I know we have `UnixfsNode` representations for them.

Reply:
conclusion here: we will eventually support Symlinks, but not for this pr. Check for symlinks & error for now.